### PR TITLE
fix(rpc/rpc): gas and gasUsed in trace root only for ParityTrace

### LINKED ISSUE #9142: Incorrect traces order for suicide call, and gas and value for create call
### Describe the bug

In this transaction: 

https://etherscan.io/tx/0xc683815f97c3de4972442448b5fa17aff31ec8f0d9e939f82234884ff989ac9e 

if we query reth:

 `curl -X POST http://IP:8545/ -H "content-type: application/json" --data '{"id":0,"params":["0x210E96",["trace"]],"method":"trace_replayBlockTransactions","jsonrpc":"2.0"}' | jq`

we get:

```
    {
      "output": "0x",
      "stateDiff": null,
      "trace": [
        {
          "type": "create",
          "action": {
            "from": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
            "gas": "0x7a120",
            "init": "0x60606040525b60006000731db3439a222c519ab44bb1144fc28167b4fa6ee6915073d8da6bf26964af9d7eed9e03e53415d37aa9604590508173ffffffffffffffffffffffffffffffffffffffff16600060023404604051809050600060405180830381858888f19350505050508073ffffffffffffffffffffffffffffffffffffffff16600060023404604051809050600060405180830381858888f19350505050503073ffffffffffffffffffffffffffffffffffffffff16ff5b5050600a8060ca6000396000f360606040526008565b00",
            "value": "0x0"
          },
          "result": {
            "address": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515",
            "code": "0x",
            "gasUsed": "0xdccf"
          },
          "subtraces": 3,
          "traceAddress": []
        },
        {
          "type": "suicide",
          "action": {
            "address": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515",
            "balance": "0x0",
            "refundAddress": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515"
          },
          "result": null,
          "subtraces": 0,
          "traceAddress": [
            2
          ]
        },
        {
          "type": "call",
          "action": {
            "from": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515",
            "callType": "call",
            "gas": "0x8fc",
            "input": "0x",
            "to": "0x1db3439a222c519ab44bb1144fc28167b4fa6ee6",
            "value": "0x58d15e17628000"
          },
          "result": {
            "gasUsed": "0x0",
            "output": "0x"
          },
          "subtraces": 0,
          "traceAddress": [
            0
          ]
        },
        {
          "type": "call",
          "action": {
            "from": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515",
            "callType": "call",
            "gas": "0x8fc",
            "input": "0x",
            "to": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
            "value": "0x58d15e17628000"
          },
          "result": {
            "gasUsed": "0x0",
            "output": "0x"
          },
          "subtraces": 0,
          "traceAddress": [
            1
          ]
        }
      ],
      "vmTrace": null,
      "transactionHash": "0xc683815f97c3de4972442448b5fa17aff31ec8f0d9e939f82234884ff989ac9e"
    },
```

suicide call comes second when it should be last, and first call has incorrect gas. Reth says first call gas is `0x7a120`, when etherscan and erigon `0x69c08`

Also value for `create` call is incorrect as well, reth says its `0x0` and erigon/etherscan `0xb1a2bc2ec50000`.

https://etherscan.io/vmtrace?txhash=0xc683815f97c3de4972442448b5fa17aff31ec8f0d9e939f82234884ff989ac9e&type=parity#raw

Erigon data:
```
 {
      "output": "0x",
      "stateDiff": null,
      "trace": [
        {
          "action": {
            "from": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
            "gas": "0x69c08",
            "init": "0x60606040525b60006000731db3439a222c519ab44bb1144fc28167b4fa6ee6915073d8da6bf26964af9d7eed9e03e53415d37aa9604590508173ffffffffffffffffffffffffffffffffffffffff16600060023404604051809050600060405180830381858888f19350505050508073ffffffffffffffffffffffffffffffffffffffff16600060023404604051809050600060405180830381858888f19350505050503073ffffffffffffffffffffffffffffffffffffffff16ff5b5050600a8060ca6000396000f360606040526008565b00",
            "value": "0xb1a2bc2ec50000"
          },
          "result": {
            "address": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515",
            "code": "0x",
            "gasUsed": "0x3577"
          },
          "subtraces": 3,
          "traceAddress": [],
          "type": "create"
        },
        {
          "action": {
            "from": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515",
            "callType": "call",
            "gas": "0x8fc",
            "input": "0x",
            "to": "0x1db3439a222c519ab44bb1144fc28167b4fa6ee6",
            "value": "0x58d15e17628000"
          },
          "result": {
            "gasUsed": "0x0",
            "output": "0x"
          },
          "subtraces": 0,
          "traceAddress": [
            0
          ],
          "type": "call"
        },
        {
          "action": {
            "from": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515",
            "callType": "call",
            "gas": "0x8fc",
            "input": "0x",
            "to": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
            "value": "0x58d15e17628000"
          },
          "result": {
            "gasUsed": "0x0",
            "output": "0x"
          },
          "subtraces": 0,
          "traceAddress": [
            1
          ],
          "type": "call"
        },
        {
          "action": {
            "address": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515",
            "refundAddress": "0xce77a9f979e700b13b8e27dd4dd9f38d383a7515",
            "balance": "0x0"
          },
          "result": null,
          "subtraces": 0,
          "traceAddress": [
            2
          ],
          "type": "suicide"
        }
      ],
      "vmTrace": null
    },
```

### Steps to reproduce

`curl -X POST http://ip:8545/ -H "content-type: application/json" --data '{"id":0,"params":["0x210E96",["trace"]],"method":"trace_replayBlockTransactions","jsonrpc":"2.0"}' | jq`

### Node logs

_No response_

### Platform(s)

Linux (x86)

### What version/commit are you on?

reth Version: 1.0.0-rc.2

### What database version are you on?

Current database version: 2
Local database version: 2

### Which chain / network are you on?

mainnet

### What type of node are you running?

Archive (default)

### What prune config do you use, if any?

_No response_

### If you've built Reth from source, provide the full command you used

_No response_

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by ZzPoLariszZ**
## Description

This pull request is Part 2/2 of fixing the bug where the `gas` and `gasUsed` fields in Parity Trace root are incorrect.

Part 1/2 paradigmxyz/revm-inspectors#171

## Related Issues and Pull Requests

- Follow: ethereum/go-ethereum#27029
- Improve: paradigmxyz/reth#3678 and paradigmxyz/reth#3719
- Fix: paradigmxyz/reth#9142 with paradigmxyz/revm-inspectors#170 
- Update: paradigmxyz/reth#3782

## Problem

The `gas` and `gasUsed` fields in Geth Debug Trace root should be the gas limit and gas used for the entire transaction.

However, two fields in Parity Trace root should be the original ones.

### Reproducible Example

With the latest version Reth v1.0.3, using `trace_transaction()` to trace  
the transaction `0x03128677ee3a9623d20f3c677f423ccc592d126374bf32e331343dd1bdf38b61`

```
curl http://localhost:8545 \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"trace_transaction","params":["0x03128677ee3a9623d20f3c677f423ccc592d126374bf32e331343dd1bdf38b61"],"id":1,"jsonrpc":"2.0"}'  
```

**From Reth**

```
gas: 0x55493 (349331)
gasUsed: 0x32d16 (208150)
```

**From [Etherscan](https://etherscan.io/vmtrace?txhash=0x03128677ee3a9623d20f3c677f423ccc592d126374bf32e331343dd1bdf38b61&type=parity#raw) and QuickNode**

```
gas: 0x4f227 (324135)
gasUsed: 0x36622 (222754)
```

## Solution for `reth`

1. Modify `revm-inspectors` part

2. Then

    - remove `with_transaction_gas_used()` for Parity Trace

    - add `with_transaction_gas_limit()` for Geth Debug Trace

## Miscellaneous

- Actually, I love the current design, but the results are inconsistent with those of others.

- When I used `make pr` to test the Reth Part, the issue paradigmxyz/reth#9381 still exists for me.

    I should only skip tests for `lockfile` and test them seperately.

### CODE REVIEW

**Comment by mattsse** [comment]
> However, two fields in Parity Trace root should be the original ones.

can you please clarify what you mean by original here?

File: crates/rpc/rpc/src/trace.rs [hunk 1]
```diff
@@ -302,15 +302,15 @@ where
                 num.into(),
                 Some(highest_idx),
                 TracingInspectorConfig::default_parity(),
-                move |tx_info, inspector, res, _, _| {
+                move |tx_info, inspector, _res, _, _| {
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:305]:
```suggestion
                move |tx_info, inspector, _, _, _| {
```

File: crates/rpc/rpc/src/trace.rs [hunk 2]
```diff
@@ -302,15 +302,15 @@ where
...
                     if let Some(idx) = tx_info.index {
                         if !indices.contains(&idx) {
                             // only record traces for relevant transactions
                             return Ok(None)
                         }
                     }
                     let traces = inspector
-                        .with_transaction_gas_used(res.gas_used())
+                        // .with_transaction_gas_used(res.gas_used())
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:313]: can remove?

File: crates/rpc/rpc/src/trace.rs [hunk 3]
```diff
@@ -368,9 +368,9 @@ where
             .spawn_trace_transaction_in_block(
                 hash,
                 TracingInspectorConfig::default_parity(),
-                move |tx_info, inspector, res, _| {
+                move |tx_info, inspector, _res, _| {
                     let traces = inspector
-                        .with_transaction_gas_used(res.result.gas_used())
+                        // .with_transaction_gas_used(res.result.gas_used())
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:373]: remove?

File: crates/rpc/rpc/src/trace.rs [hunk 4]
```diff
@@ -368,9 +368,9 @@ where
             .spawn_trace_transaction_in_block(
                 hash,
                 TracingInspectorConfig::default_parity(),
-                move |tx_info, inspector, res, _| {
+                move |tx_info, inspector, _res, _| {
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:371]:
```suggestion
                move |tx_info, inspector, _, _| {
```

File: crates/rpc/rpc/src/trace.rs [hunk 5]
```diff
@@ -387,9 +387,9 @@ where
         let traces = self.inner.eth_api.trace_block_with(
             block_id,
             TracingInspectorConfig::default_parity(),
-            |tx_info, inspector, res, _, _| {
+            |tx_info, inspector, _res, _, _| {
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:390]:
```suggestion
            |tx_info, inspector, _, _, _| {
```

File: crates/rpc/rpc/src/trace.rs [hunk 6]
```diff
@@ -387,9 +387,9 @@ where
...
                 let traces = inspector
-                    .with_transaction_gas_used(res.gas_used())
+                    // .with_transaction_gas_used(res.gas_used())
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:392]: remove?

**Review by mattsse — CHANGES_REQUESTED** [review]
style nits,
new revm-inspectors release is out

File: crates/rpc/rpc/src/debug.rs [hunk 7]
```diff
@@ -403,17 +405,17 @@ where
...
 
-        let (res, inspector) = self
+        let (res, env, inspector) = self
             .inner
             .eth_api
             .spawn_with_call_at(call, at, overrides, move |db, env| {
-                let (res, _) = this.eth_api().inspect(db, env, &mut inspector)?;
-                Ok((res, inspector))
+                let (res, _) = this.eth_api().inspect(db, env.clone(), &mut inspector)?;
+                Ok((res, env, inspector))
```
Comment (mattsse) [inline crates/rpc/rpc/src/debug.rs:413]: this clone is a bit redundant if we could return a copy of the env.tx.gas_limit instead

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]

**Review by mattsse — CHANGES_REQUESTED** [review]
can you update the revm-inspectors version in the root cargo toml, then should be okay

File: Cargo.toml [hunk 8]
```diff
@@ -386,7 +386,7 @@ revm = { version = "12.1.0", features = [
 revm-primitives = { version = "7.1.0", features = [
     "std",
 ], default-features = false }
-revm-inspectors = "0.5"
+revm-inspectors = "0.5.4"
```
Comment (mattsse) [inline Cargo.toml:389]:
```suggestion
revm-inspectors = "0.5"
```

**Review by mattsse — APPROVED** [review]

**Comment by mattsse** [comment]
merging once compiles, hehe

File: crates/rpc/rpc/src/debug.rs (hunk 7 above)
Comment (ZzPoLariszZ) [reply crates/rpc/rpc/src/debug.rs]: @mattsse It seems like the clone is still needed.

**Pull request merged**